runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkBatchPortablePipelineTranslator.java

```diff
@@ -224,10 +227,19 @@ private static void translateImpulse(
             MetricsAccumulator.getInstance());
     JavaRDD<RawUnionValue> staged = inputRdd.mapPartitions(function);
 
+    // Prevent potentially expensive re-computation of executable stage
+    SparkPipelineOptions sparkOptions =
+        context.serializablePipelineOptions.get().as(SparkPipelineOptions.class);
+    if (!sparkOptions.isCacheDisabled() && outputs.size() > 1) {
+      StorageLevel storageLevel = StorageLevel.fromString(sparkOptions.getStorageLevel());
+      staged.persist(storageLevel);
```

Comment:
It seems a bit odd that we do caching here and also implicitly on pushDataset as determined by the incrementConsumptionCount calls. 

Could we instead walk over the graph once, above, determining the total set of collections that need caching rather than spreading out the logic. (The caching itself would likely still have to happen in pushDataset.) I think this would be cleaner to follow and easier to update in the future.

Reply:
If we do this, we could also possibly compute/store the coder required for caching there too, rather than plumbing it all over.